pydantic_ai_slim/pydantic_ai/agent.py

```diff
@@ -421,7 +422,9 @@ def __init__(
             self._output_toolset.max_retries = self._max_result_retries
 
         self._function_toolset = FunctionToolset(tools, max_retries=retries)
-        self._user_toolsets = toolsets or ()
+        self._user_toolsets = [
+            toolset if isinstance(toolset, AbstractToolset) else DynamicToolset(toolset) for toolset in toolsets or []
```

Comment:
Whoops, this `DynamicToolset` would be shared across multiple uses of the agent (possibly with different deps for different users), so this won't work right. We're better off storing the functions as you had it, and then building the `DynamicToolset` inside `iter` (as you had it). And making `DynamicToolset` private as we don't actually want people to create it themselves and run into this issue.

Or we'd need to fix `DynamicToolset` to keep a separate toolset per deps, but then deps would need to be hashable, which is not something we want to require.

Reply:
@strawgate I think it'll work if we only build the `DynamicToolset` inside `iter` and pass it to `ToolManager` there (via the `_get_toolset` call as you were doing before), so every run will get its own toolset.

Reply:
I looked at this a bit and I'm wondering if the simplest solution is to just add a copy method to the dynamic toolset and just make a copy of it in iter / _get_toolset

will make a commit for this to look at